src/select.cpp

```
@@ -32,43 +32,29 @@ SEXP select_not_grouped(const DataFrame& df, const SymbolVector& keep, const Sym
 }
 
 DataFrame select_grouped(GroupedDataFrame gdf, const SymbolVector& keep, const SymbolVector& new_names) {
+  // start by selecting the columns without taking care of the grouping structure
   DataFrame copy = select_not_grouped(gdf.data(), keep, new_names);
 
-  SymbolMap keep_map(keep);
+  // then handle the groups attribute
+  // it is almost the same as the groups attribute of the input data frame, but
+  // names might change so we need to create a shallow copy and then deal with the names
+  DataFrame groups(shallow_copy(List(gdf.group_data())));
 
-  // handle vars  attribute : make a shallow copy of the list and alter
-  //   its names attribute
-  SymbolVector vars(get_vars(copy, true));
-
-  int nv = vars.size();
-  for (int i = 0; i < nv; i++) {
-    SymbolString s = vars[i];
-    SymbolMapIndex j = keep_map.get_index(s);
-    if (j.origin != NEW) {
-      vars.set(i, new_names[j.pos]);
+  // update the names of the grouping variables in case they are involved in
+  // the selection, i.e. select(data, g1 = g2)
+  CharacterVector group_names = clone<CharacterVector>(groups.names());
+  IntegerVector positions = keep.match(group_names);
+  int nl = group_names.size();
+  for (int i = 0; i < nl; i++) {
+    int pos = positions[i];
+    if (pos != NA_INTEGER) {
```

Do we need to throw an error in case of mismatch?

---

no because this would have been caught much earlier by `vars_select`. I guess that test is useless.

---

The test saves us from a segmentation fault if for some reason the group variables are omitted. Can we just `stop()` if this test hits?

---

Do you have code that used to trigger these segfaults. I don't see how pos can ever be NA here.

---

I don't have code now. But this code is at the mercy of two other components (R code in *dplyr*, and *tidyselect*) working correctly. If they for some reason fail, a `stop()` would give us a very clear indication about the problem. A segfault is just as clear (for the developer), but it tends to scare users.

Also, doesn't `group_data()` contain `.rows` as the last column?

---

There actually was a test which I had to update for this new error: 

```
test_that("rename errors with invalid grouped data frame (#640)", {
  df <- data_frame(a = 1:3, b = 2:4, d = 3:5) %>% group_by(a, b)
  df$a <- NULL
  expect_error(
    df %>% rename(e = d),
    "not found in groups metadata"
  )
  expect_error(
    df %>% rename(e = b),
    "not found in groups metadata"
  )
})
```

---

🤦‍♂️ except that I pushed it to the wrong branch `nest_join-3570` that depends on this one. I'll just merge this one now and the other one quickly after. Need more Monday ☕️.